{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "parser = argparse.ArgumentParser(\n",
    "                    prog='Training models using Machine Learning',\n",
    "                    description='Logistic Regression, Decision Tree, and SVM',\n",
    "                    epilog='Help')\n",
    "\n",
    "parser.add_argument('input', metavar='i', type=str, help='Input path')\n",
    "parser.add_argument('output', metavar='o', type=str, help='Outphut path')\n",
    "args = parser.parse_args()\n",
    "'''\n",
    "\n",
    "#vcf_path = args.input\n",
    "vcf_path = \"../data/HLA/HLA1_chr6.vcf\"\n",
    "# Read a vcf file\n",
    "def read_vcf(vcf_path):\n",
    "    with open(vcf_path, \"rt\") as ifile:\n",
    "          for line in ifile:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                  vcf_names = [x for x in line.split('\\t')]\n",
    "                  break\n",
    "    ifile.close()\n",
    "    data = pd.read_csv(vcf_path, comment='#', delim_whitespace=True, header=None, names=vcf_names)\n",
    "    return data\n",
    "data = read_vcf(vcf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno = data.iloc[:, 9::]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binary_sampler(p, rows, cols):\n",
    "  '''Sample binary random variables.\n",
    "  \n",
    "  Args:\n",
    "    - p: probability of 1\n",
    "    - rows: the number of rows\n",
    "    - cols: the number of columns\n",
    "    \n",
    "  Returns:\n",
    "    - binary_random_matrix: generated binary random matrix.\n",
    "  '''\n",
    "  np.random.seed(7)\n",
    "  unif_random_matrix = np.random.uniform(0., 1., size = [rows, cols])\n",
    "  binary_random_matrix = 1*(unif_random_matrix < p)\n",
    "  return binary_random_matrix\n",
    "\n",
    "def get_data(X, miss_rate):\n",
    "    # Parameters\n",
    "    no, dim = X.shape\n",
    "\n",
    "    # Introduce missing data\n",
    "    data_m = binary_sampler(1-miss_rate, no, dim)\n",
    "    miss_data_x = X.copy()\n",
    "    miss_data_x[data_m == 0] = \".|.\"\n",
    "    return X, miss_data_x, data_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno, geno_miss, m = get_data(geno, 0.1)\n",
    "gn = pd.concat([data.iloc[:, 0:9], geno_miss], axis = 1)\n",
    "gn.rename(columns={'NA21144\\n':'NA21144'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = \"\"\"##fileformat=VCFv4.1\n",
    "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
    "##fileDate=20150218\n",
    "##reference=ftp://ftp.1000genomes.ebi.ac.uk//vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz\n",
    "##source=1000GenomesPhase3Pipeline\n",
    "##contig=<ID=chr1,assembly=b37,length=249250621>\n",
    "##contig=<ID=chr2,assembly=b37,length=243199373>\n",
    "##contig=<ID=chr3,assembly=b37,length=198022430>\n",
    "##contig=<ID=chr4,assembly=b37,length=191154276>\n",
    "##contig=<ID=chr5,assembly=b37,length=180915260>\n",
    "##contig=<ID=chr6,assembly=b37,length=171115067>\n",
    "##contig=<ID=chr7,assembly=b37,length=159138663>\n",
    "##contig=<ID=chr8,assembly=b37,length=146364022>\n",
    "##contig=<ID=chr9,assembly=b37,length=141213431>\n",
    "##contig=<ID=chr10,assembly=b37,length=135534747>\n",
    "##contig=<ID=chr11,assembly=b37,length=135006516>\n",
    "##contig=<ID=chr12,assembly=b37,length=133851895>\n",
    "##contig=<ID=chr13,assembly=b37,length=115169878>\n",
    "##contig=<ID=chr14,assembly=b37,length=107349540>\n",
    "##contig=<ID=chr15,assembly=b37,length=102531392>\n",
    "##contig=<ID=chr16,assembly=b37,length=90354753>\n",
    "##contig=<ID=chr17,assembly=b37,length=81195210>\n",
    "##contig=<ID=chr18,assembly=b37,length=78077248>\n",
    "##contig=<ID=chr19,assembly=b37,length=59128983>\n",
    "##contig=<ID=chr20,assembly=b37,length=63025520>\n",
    "##contig=<ID=chr21,assembly=b37,length=48129895>\n",
    "##contig=<ID=chr22,assembly=b37,length=51304566>\n",
    "##contig=<ID=GL000191.1,assembly=b37,length=106433>\n",
    "##contig=<ID=GL000192.1,assembly=b37,length=547496>\n",
    "##contig=<ID=GL000193.1,assembly=b37,length=189789>\n",
    "##contig=<ID=GL000194.1,assembly=b37,length=191469>\n",
    "##contig=<ID=GL000195.1,assembly=b37,length=182896>\n",
    "##contig=<ID=GL000196.1,assembly=b37,length=38914>\n",
    "##contig=<ID=GL000197.1,assembly=b37,length=37175>\n",
    "##contig=<ID=GL000198.1,assembly=b37,length=90085>\n",
    "##contig=<ID=GL000199.1,assembly=b37,length=169874>\n",
    "##contig=<ID=GL000200.1,assembly=b37,length=187035>\n",
    "##contig=<ID=GL000201.1,assembly=b37,length=36148>\n",
    "##contig=<ID=GL000202.1,assembly=b37,length=40103>\n",
    "##contig=<ID=GL000203.1,assembly=b37,length=37498>\n",
    "##contig=<ID=GL000204.1,assembly=b37,length=81310>\n",
    "##contig=<ID=GL000205.1,assembly=b37,length=174588>\n",
    "##contig=<ID=GL000206.1,assembly=b37,length=41001>\n",
    "##contig=<ID=GL000207.1,assembly=b37,length=4262>\n",
    "##contig=<ID=GL000208.1,assembly=b37,length=92689>\n",
    "##contig=<ID=GL000209.1,assembly=b37,length=159169>\n",
    "##contig=<ID=GL000210.1,assembly=b37,length=27682>\n",
    "##contig=<ID=GL000211.1,assembly=b37,length=166566>\n",
    "##contig=<ID=GL000212.1,assembly=b37,length=186858>\n",
    "##contig=<ID=GL000213.1,assembly=b37,length=164239>\n",
    "##contig=<ID=GL000214.1,assembly=b37,length=137718>\n",
    "##contig=<ID=GL000215.1,assembly=b37,length=172545>\n",
    "##contig=<ID=GL000216.1,assembly=b37,length=172294>\n",
    "##contig=<ID=GL000217.1,assembly=b37,length=172149>\n",
    "##contig=<ID=GL000218.1,assembly=b37,length=161147>\n",
    "##contig=<ID=GL000219.1,assembly=b37,length=179198>\n",
    "##contig=<ID=GL000220.1,assembly=b37,length=161802>\n",
    "##contig=<ID=GL000221.1,assembly=b37,length=155397>\n",
    "##contig=<ID=GL000222.1,assembly=b37,length=186861>\n",
    "##contig=<ID=GL000223.1,assembly=b37,length=180455>\n",
    "##contig=<ID=GL000224.1,assembly=b37,length=179693>\n",
    "##contig=<ID=GL000225.1,assembly=b37,length=211173>\n",
    "##contig=<ID=GL000226.1,assembly=b37,length=15008>\n",
    "##contig=<ID=GL000227.1,assembly=b37,length=128374>\n",
    "##contig=<ID=GL000228.1,assembly=b37,length=129120>\n",
    "##contig=<ID=GL000229.1,assembly=b37,length=19913>\n",
    "##contig=<ID=GL000230.1,assembly=b37,length=43691>\n",
    "##contig=<ID=GL000231.1,assembly=b37,length=27386>\n",
    "##contig=<ID=GL000232.1,assembly=b37,length=40652>\n",
    "##contig=<ID=GL000233.1,assembly=b37,length=45941>\n",
    "##contig=<ID=GL000234.1,assembly=b37,length=40531>\n",
    "##contig=<ID=GL000235.1,assembly=b37,length=34474>\n",
    "##contig=<ID=GL000236.1,assembly=b37,length=41934>\n",
    "##contig=<ID=GL000237.1,assembly=b37,length=45867>\n",
    "##contig=<ID=GL000238.1,assembly=b37,length=39939>\n",
    "##contig=<ID=GL000239.1,assembly=b37,length=33824>\n",
    "##contig=<ID=GL000240.1,assembly=b37,length=41933>\n",
    "##contig=<ID=GL000241.1,assembly=b37,length=42152>\n",
    "##contig=<ID=GL000242.1,assembly=b37,length=43523>\n",
    "##contig=<ID=GL000243.1,assembly=b37,length=43341>\n",
    "##contig=<ID=GL000244.1,assembly=b37,length=39929>\n",
    "##contig=<ID=GL000245.1,assembly=b37,length=36651>\n",
    "##contig=<ID=GL000246.1,assembly=b37,length=38154>\n",
    "##contig=<ID=GL000247.1,assembly=b37,length=36422>\n",
    "##contig=<ID=GL000248.1,assembly=b37,length=39786>\n",
    "##contig=<ID=GL000249.1,assembly=b37,length=38502>\n",
    "##contig=<ID=MT,assembly=b37,length=16569>\n",
    "##contig=<ID=NC_007605,assembly=b37,length=171823>\n",
    "##contig=<ID=chrX,assembly=b37,length=155270560>\n",
    "##contig=<ID=Y,assembly=b37,length=59373566>\n",
    "##contig=<ID=hs37d5,assembly=b37,length=35477943>\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##INFO=<ID=CIEND,Number=2,Type=Integer,Description=\"Confidence interval around END for imprecise variants\">\n",
    "##INFO=<ID=CIPOS,Number=2,Type=Integer,Description=\"Confidence interval around POS for imprecise variants\">\n",
    "##INFO=<ID=CS,Number=1,Type=String,Description=\"Source call set.\">\n",
    "##INFO=<ID=END,Number=1,Type=Integer,Description=\"End coordinate of this variant\">\n",
    "##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description=\"Imprecise structural variation\">\n",
    "##INFO=<ID=MC,Number=.,Type=String,Description=\"Merged calls.\">\n",
    "##INFO=<ID=MEINFO,Number=4,Type=String,Description=\"Mobile element info of the form NAME,START,END<POLARITY; If there is only 5' OR 3' support for this call, will be NULL NULL for START and END\">\n",
    "##INFO=<ID=MEND,Number=1,Type=Integer,Description=\"Mitochondrial end coordinate of inserted sequence\">\n",
    "##INFO=<ID=MLEN,Number=1,Type=Integer,Description=\"Estimated length of mitochondrial insert\">\n",
    "##INFO=<ID=MSTART,Number=1,Type=Integer,Description=\"Mitochondrial start coordinate of inserted sequence\">\n",
    "##INFO=<ID=SVLEN,Number=.,Type=Integer,Description=\"SV length. It is only calculated for structural variation MEIs. For other types of SVs; one may calculate the SV length by INFO:END-START+1, or by finding the difference between lengthes of REF and ALT alleles\">\n",
    "##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">\n",
    "##INFO=<ID=TSD,Number=1,Type=String,Description=\"Precise Target Site Duplication for bases, if unknown, value will be NULL\">\n",
    "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Total number of alternate alleles in called genotypes\">\n",
    "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1)\">\n",
    "##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of samples with data\">\n",
    "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
    "##INFO=<ID=EAS_AF,Number=A,Type=Float,Description=\"Allele frequency in the EAS populations calculated from AC and AN, in the range (0,1)\">\n",
    "##INFO=<ID=EUR_AF,Number=A,Type=Float,Description=\"Allele frequency in the EUR populations calculated from AC and AN, in the range (0,1)\">\n",
    "##INFO=<ID=AFR_AF,Number=A,Type=Float,Description=\"Allele frequency in the AFR populations calculated from AC and AN, in the range (0,1)\">\n",
    "##INFO=<ID=AMR_AF,Number=A,Type=Float,Description=\"Allele frequency in the AMR populations calculated from AC and AN, in the range (0,1)\">\n",
    "##INFO=<ID=SAS_AF,Number=A,Type=Float,Description=\"Allele frequency in the SAS populations calculated from AC and AN, in the range (0,1)\">\n",
    "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total read depth; only low coverage data were counted towards the DP, exome data were not used\">\n",
    "##INFO=<ID=AA,Number=1,Type=String,Description=\"Ancestral Allele. Format: AA|REF|ALT|IndelType. AA: Ancestral allele, REF:Reference Allele, ALT:Alternate Allele, IndelType:Type of Indel (REF, ALT and IndelType are only defined for indels)\">\n",
    "##INFO=<ID=VT,Number=.,Type=String,Description=\"indicates what type of variant the line represents\">\n",
    "##INFO=<ID=EX_TARGET,Number=0,Type=Flag,Description=\"indicates whether a variant is within the exon pull down target boundaries\">\n",
    "##INFO=<ID=MULTI_ALLELIC,Number=0,Type=Flag,Description=\"indicates whether a site is multi-allelic\">\n",
    "##INFO=<ID=STRAND_FLIP,Number=0,Type=Flag,Description=\"Indicates that the reference strand has changed between GRCh37 and GRCh38\">\n",
    "##INFO=<ID=REF_SWITCH,Number=0,Type=Flag,Description=\"Indicates that the reference allele has changed\">\n",
    "##INFO=<ID=DEPRECATED_RSID,Number=.,Type=String,Description=\"dbsnp rs IDs that have been merged into other rs IDs or do not map to GRCh38\">\n",
    "##INFO=<ID=RSID_REMOVED,Number=.,Type=String,Description=\"dbsnp rs IDs removed from this variant, due to either the variant splitting up or being deprecated/merged\">\n",
    "##INFO=<ID=GRCH37_38_REF_STRING_MATCH,Number=0,Type=Flag,Description=\"Indicates reference allele in origin GRCh37 vcf string-matches reference allele in dbsnp GRCh38 vcf\">\n",
    "##INFO=<ID=NOT_ALL_RSIDS_STRAND_CHANGE_OR_REF_SWITCH,Number=0,Type=Flag,Description=\"Indicates only some of the rs IDs in origin GRCh37 vcf switched strands or switched strands and changed reference allele. This would result in rs IDs being split into multiple lines\">\n",
    "##INFO=<ID=GRCH37_POS,Number=1,Type=Integer,Description=\"Position in origin GRCh37 vcf\">\n",
    "##INFO=<ID=GRCH37_REF,Number=1,Type=String,Description=\"Representation of reference allele in origin GRCh37 vcf\">\n",
    "##INFO=<ID=ALLELE_TRANSFORM,Number=0,Type=Flag,Description=\"Indicates that at least some of the alleles have changed in how they're represented, e.g. through left shifting.\">\n",
    "##INFO=<ID=REF_NEW_ALLELE,Number=0,Type=Flag,Description=\"Indicates that the reference allele is an allele not present in the origin GRCh37 vcf\">\n",
    "##INFO=<ID=CHROM_CHANGE_BETWEEN_ASSEMBLIES,Number=.,Type=String,Description=\"dbsnp rs IDs that are mapped to a different chromosome between GRCh37 and GRCh38\">\n",
    "##bcftools_annotateVersion=1.9+htslib-1.9\n",
    "##bcftools_annotateCommand=annotate --rename-chrs chr_names.txt -Ou ALL.chr6_GRCh38.genotypes.20170504.vcf.gz; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_viewVersion=1.9+htslib-1.9\n",
    "##bcftools_viewCommand=view -e 'INFO/AC<3 | INFO/AN-INFO/AC<3' -Ou; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_normVersion=1.9+htslib-1.9\n",
    "##bcftools_normCommand=norm -m -any -Ou; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_viewCommand=view -i 'INFO/VT=\"SNP\" | INFO/VT=\"INDEL\"' -Ou; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_normCommand=norm -f ../ref_genome/hg38.fasta -d none -Ou; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_viewCommand=view -m 2 -M 2 -Ou; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_viewCommand=view -g ^miss -Oz -o 1000GP_chr6.vcf.gz; Date=Mon Aug 28 20:38:30 2023\n",
    "##bcftools_viewCommand=view -e ID=@1000GP_chr6.dup_id -Oz -o 1000GP_filtered_chr6.vcf.gz 1000GP_chr6.vcf.gz; Date=Tue Aug 29 18:58:01 2023\n",
    "##bcftools_pluginVersion=1.9+htslib-1.9\n",
    "##bcftools_pluginCommand=plugin fill-tags -Oz -o 1000GP_AF_chr6.vcf.gz -- 1000GP_filtered_chr6.vcf.gz -t AF; Date=Sun Sep 24 18:38:53 2023\n",
    "##bcftools_viewVersion=1.14+htslib-1.15.1\n",
    "##bcftools_viewCommand=view -R HLA1_chr6.bed 1000GP_AF_chr6.vcf.gz; Date=Thu Mar 28 08:24:00 2024\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_VCF = \"../data/HLA/hla.vcf\"\n",
    "with open(output_VCF, 'w') as vcf:\n",
    "    vcf.write(header)\n",
    "gn.to_csv(output_VCF, sep=\"\\t\", mode='a', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(columns = list(gn.columns)[9::], data = m).to_csv(\"../data/HLA/mask.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "imputation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
